XenAdmin/Wizards/PatchingWizard/PlanActions/ApplyXenServerPatchPlanAction.cs

```diff
@@ -62,13 +62,25 @@ protected override void RunWithSession(ref Session session)
 
             if (mapping != null && (mapping.Pool_patch != null || mapping.Pool_update != null))
             {
-                AddProgressStep(string.Format(Messages.UPDATES_WIZARD_APPLYING_UPDATE, xenServerPatch.Name, host.Name()));
+                try
+                {
+                    AddProgressStep(string.Format(Messages.UPDATES_WIZARD_APPLYING_UPDATE, xenServerPatch.Name,
+                        host.Name()));
 
-                var task = mapping.Pool_patch == null
-                    ? Pool_update.async_apply(session, mapping.Pool_update.opaque_ref, host.opaque_ref)
-                    : Pool_patch.async_apply(session, mapping.Pool_patch.opaque_ref, host.opaque_ref);
+                    var task = mapping.Pool_patch == null
+                        ? Pool_update.async_apply(session, mapping.Pool_update.opaque_ref, host.opaque_ref)
+                        : Pool_patch.async_apply(session, mapping.Pool_patch.opaque_ref, host.opaque_ref);
 
-                PollTaskForResultAndDestroy(Connection, ref session, task);
+                    PollTaskForResultAndDestroy(Connection, ref session, task);
+                }
+                catch (Failure f)
+                {
+                    if (f.ErrorDescription.Count > 1 && (f.ErrorDescription[0] == Failure.PATCH_ALREADY_APPLIED || f.ErrorDescription[0] == Failure.UPDATE_ALREADY_APPLIED))

```

Comment:
Might it be useful to also AddProgressStep("patch already applied")? Asking because we show the user again the applying update step in line 67.

Reply:
Shall we do a ReplaceProgressStep? to replace "Installing update" with "already applied"?

Reply:
Yes, this sounds much better.
Irrelevant to this PR, but it's just occurred to me that we should probably be doing ReplaceProgress here too: https://github.com/xenserver/xenadmin/blob/REQ-672/XenAdmin/Wizards/PatchingWizard/PlanActions/DownloadPatchPlanAction.cs#L73